### PR TITLE
TM-627: Refactor code Search

### PR DESCRIPTION
**Pull request by ManhNTX**

### CODE REVIEW

File: lib/main/localizations/app_localizations.dart [hunk 1]
```diff
@@ -1453,4 +1453,88 @@ class AppLocalizations {
     return Intl.message('Server URL is not valid, please try again',
         name: 'wrongUrlMessage');
   }
+
```
Comment (hoangdat) [inline lib/main/localizations/app_localizations.dart:1456]: Please generate intl_messages with new string term

File: core/lib/utils/wrapper.dart [hunk 2]
```diff
@@ -0,0 +1,4 @@
+class Wrapped<T> {
```
Comment (hoangdat) [inline core/lib/utils/wrapper.dart:1]: Can you explain about this? Quite cool?

File: lib/features/mailbox_dashboard/presentation/controller/search_controller.dart [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+  FocusNode searchFocus = FocusNode();
+
+  SearchEmailFilter get searchEmailFilterValue => searchEmailFilter.value;
+
+  SearchController(
+      this._quickSearchEmailInteractor,
+      );
+
+  selectOpenAdvanceSearch() {
+    isAdvancedSearchViewOpen.value = !isAdvancedSearchViewOpen.value;
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/controller/search_controller.dart:48]: you should using  function `toggle()` of `GetX` for object `boolean obx`
Ex: `isAdvancedSearchViewOpen.toggle();`

File: core/lib/utils/wrapper.dart (hunk 2 above)
Comment (ManhNTX) [reply core/lib/utils/wrapper.dart:1]: @hoangdat this class uses send parameter is null in function copyWith

File: lib/features/mailbox_dashboard/presentation/controller/search_controller.dart [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+  void registerSearchFocusListener() {
+    searchFocus.addListener(() {
+      final hasFocus = searchFocus.hasFocus;
+      final query = searchEmailFilter.value.text?.value;
+      log('MailboxDashBoardController::_registerSearchFocusListener(): hasFocus: $hasFocus | query: $query');
+      if (!hasFocus && (query == null || query.isEmpty)) {
+        updateFilterEmail(text: SearchQuery.initial());
+        searchInputController.clear();
+        cleanSearchFilter();
+        searchFocus.unfocus();      }
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/controller/search_controller.dart:132]: newline for the `}` character.

File: lib/features/mailbox_dashboard/presentation/model/search/quick_search_filter.dart [hunk 5]
```diff
@@ -43,20 +39,18 @@ extension QuickSearchFilterExtension on QuickSearchFilter {
     }
   }
 
-  Color getBackgroundColor(List<QuickSearchFilter> listFilter) {
-    final filterSelected = listFilter.contains(this);
+  Color getBackgroundColor(bool quickSearchFilterSelected) {
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/model/search/quick_search_filter.dart:42]: please use 
```
Color getBackgroundColor({bool quickSearchFilterSelected})
```

File: lib/features/mailbox_dashboard/presentation/model/search/quick_search_filter.dart [hunk 6]
```diff
@@ -26,10 +23,9 @@ extension QuickSearchFilterExtension on QuickSearchFilter {
     }
   }
 
-  String getIcon(ImagePaths imagePaths, List<QuickSearchFilter> listFilter) {
-    final filterSelected = listFilter.contains(this);
+  String getIcon(ImagePaths imagePaths, bool quickSearchFilterSelected) {
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/model/search/quick_search_filter.dart:26]:
```
String getIcon(ImagePaths imagePaths, {bool quickSearchFilterSelected})
```

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 7]
```diff
@@ -225,19 +198,9 @@ class MailboxDashBoardController extends ReloadableController {
...
-      log('MailboxDashBoardController::_registerSearchFocusListener(): hasFocus: $hasFocus | query: $query');
-      if (!hasFocus && (query == null || query.isEmpty)) {
-        disableSearch();
-      }
-    });
-  }
-
   void _getUserProfile() async {
     userProfile.value = sessionCurrent != null ? UserProfile(sessionCurrent!.username.value) : null;
+    searchController.userProfile.value = userProfile.value;
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:203]: I think we accepted: we don't follow this way

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 8]
```diff
@@ -246,6 +209,7 @@ class MailboxDashBoardController extends ReloadableController {
     if (arguments is Session) {
       sessionCurrent = arguments;
       accountId.value = sessionCurrent?.accounts.keys.first;
+      searchController.accountId.value = accountId.value;
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:212]: idem

File: lib/features/mailbox_dashboard/presentation/controller/search_controller.dart [hunk 9]
```diff
@@ -179,6 +194,19 @@ class SearchController extends BaseController {
     }
   }
 
+  void saveRecentSearch(RecentSearch recentSearch) {
+    consumeState(_saveRecentSearchInteractor.execute(recentSearch));
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/search_controller.dart:199]: `states` come from this interactor will be handle in where?

Comment (ManhNTX) [reply lib/features/mailbox_dashboard/presentation/controller/search_controller.dart:200]: `states ` will not handle (or will be handle in future). This code created on mailbox_dashboard_controller before, I just move it to search controller.

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart (hunk 7 above)
Comment (ManhNTX) [reply lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart]: It seems like search_controller will binding before mailbox_dashboard_controller.dart (you can check it in [here](lib/features/mailbox_dashboard/presentation/mailbox_dashboard_bindings.dart) ). So we can't use this way talk before. Now I added `userProfile` and `accountId` as parameter of functions. when we use functions just add it as parameter.

Comment (hoangdat) [reply lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart]: Why we need to binding `SearchController` before `mailbox_dashboard_controller`? What is your strategy when separate the `controllers`?

File: lib/features/mailbox_dashboard/presentation/controller/search_controller.dart [hunk 10]
```diff
@@ -70,32 +64,40 @@ class SearchController extends BaseController {
 
   void selectQuickSearchFilter({
     required QuickSearchFilter quickSearchFilter,
+    required UserProfile userProfile,
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/search_controller.dart:67]: cool! We should think about `decouple SearchController to any controller`

File: lib/l10n/intl_messages.arb [hunk 11]
```diff
@@ -1,5 +1,5 @@
 {
-  "@@last_modified": "2022-06-24T14:16:53.170960",
+  "@@last_modified": "2022-06-25T23:25:44.320884",
```
Comment (hoangdat) [inline lib/l10n/intl_messages.arb:2]: no need this change

File: test/features/thread/domain/usecases/get_emails_in_mailbox_interactor_test.mocks.dart [hunk 12]
```diff
@@ -1,4 +1,4 @@
-// Mocks generated by Mockito 5.0.17 from annotations
+// Mocks generated by Mockito 5.2.0 from annotations
```
Comment (hoangdat) [inline test/features/thread/domain/usecases/get_emails_in_mailbox_interactor_test.mocks.dart:1]: no need this file's changes

File: test/features/thread/domain/usecases/refresh_changes_emails_in_mailbox_interactor_test.mocks.dart [hunk 13]
```diff
@@ -1,4 +1,4 @@
-// Mocks generated by Mockito 5.0.17 from annotations
+// Mocks generated by Mockito 5.2.0 from annotations
```
Comment (hoangdat) [inline test/features/thread/domain/usecases/refresh_changes_emails_in_mailbox_interactor_test.mocks.dart:1]: idem

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart (hunk 7 above)
Comment (ManhNTX) [reply lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart]: I think you had an answer ( like below https://github.com/linagora/tmail-flutter/pull/683#discussion_r906734865). In the future, I want to move feature search to 1 module ( easy to re-use and maintain ). So  I think we should let other controllers depend on the searchcontroller, not the other way around.

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 14]
```diff
@@ -298,53 +261,29 @@ class MailboxDashBoardController extends ReloadableController {
...
-  }
-
-  void updateTextSearch(String value) {
-    searchInputController.text = value;
-  }
-
-  void searchEmail(BuildContext context, String value) {
-    searchQuery = SearchQuery(value);
-    dispatchState(Right(SearchEmailNewQuery(searchQuery ?? SearchQuery.initial())));
+  void searchEmail(BuildContext context,String value) {
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:267]: There should be a space character between the 2 parameters. `searchEmail(BuildContext context, String value)`

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 15]
```diff
@@ -298,53 +261,29 @@ class MailboxDashBoardController extends ReloadableController {
...
+    searchController.updateFilterEmail(text: SearchQuery(value));
+    dispatchState(Right(SearchEmailNewQuery(searchController.searchEmailFilter.value.text ?? SearchQuery.initial())));
     FocusScope.of(context).unfocus();
-    if (searchInsideEmailDetailedViewIsActive(context)) {
-      closeEmailDetailedView();
+    if (_searchInsideEmailDetailedViewIsActive(context)) {
+      _closeEmailDetailedView();
+    }
+    if(value.isEmpty){
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:274]: same `if (value.isEmpty) {`

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart [hunk 16]
```diff
@@ -150,21 +150,21 @@ class MailboxDashBoardView extends BaseMailboxDashBoardView {
         if (controller.isSelectionEnabled()) {
           return _buildListButtonTopBarSelection(context);
         } else {
-          return controller.isSearchActive()
+          return controller.searchController.isSearchActive()
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart:153]: IMO, You can create a searchController object in the MailboxDashBoardView. 
Ex: `final searchController = Get.find<SearchController>();` for reusability, 
Avoid calling `controller.searchController` repeatedly which messes up the code

File: lib/features/thread/presentation/thread_controller.dart [hunk 17]
```diff
@@ -280,8 +282,8 @@ class ThreadController extends BaseController {
       if (viewState is Either) {
         viewState.map((success) {
           if (success is SearchEmailNewQuery){
-            mailboxDashBoardController.clearState();
             _searchEmail();
+            mailboxDashBoardController.clearState();
```
Comment (dab246) [inline lib/features/thread/presentation/thread_controller.dart:286]: Why leave `clearState` after `_searchEmail `?

**Review by dab246 — APPROVED** [review]

**Review by hoangdat — APPROVED** [review]

**Pull request merged**